File: scripts/fetch-guofeng-images.mjs
#!/usr/bin/env node

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const API = "https://commons.wikimedia.org/w/api.php";
const UA = "itbxm-guofeng-image-fetcher/1.0 (skill asset workflow)";

export function parseArgs(argv) {
  const parsed = {
    query: [],
    count: 2,
    width: 1600,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];

    if (arg === "--query") {
      i += 1;
      while (i < argv.length && !argv[i].startsWith("--")) {
        parsed.query.push(argv[i]);
        i += 1;
      }
      i -= 1;
      continue;
    }

    if (arg === "--out") {
      parsed.out = argv[++i];
      continue;
    }

    if (arg === "--count") {
      parsed.count = Number.parseInt(argv[++i], 10);
      continue;
    }

    if (arg === "--width") {
      parsed.width = Number.parseInt(argv[++i], 10);
      continue;
    }

    if (arg === "--help" || arg === "-h") {
      parsed.help = true;
      continue;
    }

    throw new Error(`Unknown argument: ${arg}`);
  }

  if (!parsed.help) {
    if (parsed.query.length === 0) throw new Error("Missing --query");
    if (!parsed.out) throw new Error("Missing --out");
    if (!Number.isInteger(parsed.count) || parsed.count < 1) throw new Error("--count must be a positive integer");
    if (!Number.isInteger(parsed.width) || parsed.width < 200) throw new Error("--width must be an integer >= 200");
  }

  return parsed;
}

export function safeFileName(value) {
  return String(value)
    .normalize("NFKC")
    .replace(/[^\p{L}\p{N}]+/gu, "_")
    .replace(/^_+|_+$/g, "")
    .slice(0, 80) || "image";
}

export function buildSourceEntry({ localFile, query, sourceUrl, license, artist, retrievedAt }) {
  return `| ${localFile} | ${query} | ${sourceUrl} | ${license || "?"} | ${artist || "?"} | ${retrievedAt} |\n`;
}

function stripHtml(value) {
  return String(value || "?")
    .replace(/<[^>]+>/g, "")
    .replace(/\s+/g, " ")
    .trim() || "?";
}

function imageExtension(url) {
  try {
    const pathname = new URL(url).pathname;
    const ext = path.extname(pathname).toLowerCase();
    if (/^\.(?:jpg|jpeg|png|webp|gif|svg)$/.test(ext)) return ext;
  } catch {
    // Fall through to default.
  }
  return ".jpg";
}

function sourcesPathForOut(outDir) {
  const absoluteOut = path.resolve(outDir);
  const parent = path.dirname(absoluteOut);
  const sourcesDir = path.basename(parent) === "assets"
    ? path.join(parent, "sources")
    : path.join(path.dirname(absoluteOut), "sources");
  return path.join(sourcesDir, "image-sources.md");
}

function siteRootForOut(outDir) {
  const absoluteOut = path.resolve(outDir);
  const parent = path.dirname(absoluteOut);
  if (path.basename(parent) === "assets") return path.dirname(parent);
  return process.cwd();
}

async function apiGet(params) {
  const url = `${API}?${new URLSearchParams(params).toString()}`;
  const response = await fetch(url, {
    headers: {
      "User-Agent": UA,
      Accept: "application/json",
    },
  });

  if (!response.ok) {
    throw new Error(`Wikimedia API returned ${response.status} ${response.statusText}`);
  }

  return response.json();
}

async function downloadFile(url, filePath) {
  const response = await fetch(url, {
    headers: {
      "User-Agent": UA,
    },
  });

  if (!response.ok) {
    throw new Error(`download returned ${response.status} ${response.statusText}`);
  }

  const buffer = Buffer.from(await response.arrayBuffer());
  await fs.promises.writeFile(filePath, buffer);
}

async function fetchQuery(query, options) {
  const data = await apiGet({
    action: "query",
    format: "json",
    generator: "search",
    gsrsearch: query,
    gsrnamespace: "6",
    gsrlimit: String(options.count),
    prop: "imageinfo",
    iiprop: "url|extmetadata",
    iiurlwidth: String(options.width),
  });

  const pages = Object.values(data.query?.pages || {});
  const records = [];

  for (const page of pages.slice(0, options.count)) {
    const imageInfo = page.imageinfo?.[0];
    const imageUrl = imageInfo?.thumburl || imageInfo?.url;
    if (!imageUrl) continue;

    const meta = imageInfo.extmetadata || {};
    const license = stripHtml(meta.LicenseShortName?.value);
    const artist = stripHtml(meta.Artist?.value);
    const sourceUrl = imageInfo.descriptionurl || "";
    const title = String(page.title || "image").replace(/^File:/, "");
    const ext = imageExtension(imageUrl);
    const fileName = `${safeFileName(`${query}_${title}`)}${ext}`;
    const filePath = path.join(options.out, fileName);

    await downloadFile(imageUrl, filePath);

    records.push({
      localFile: path.relative(options.siteRoot, filePath).replaceAll(path.sep, "/"),
      query,
      sourceUrl,
      license,
      artist,
    });
  }

  return records;
}

async function appendSources(sourceFile, records, retrievedAt) {
  await fs.promises.mkdir(path.dirname(sourceFile), { recursive: true });

  let content = "";
  if (!fs.existsSync(sourceFile)) {
    content += "# Image Sources\n\n";
    content += "| Local file | Query | Source URL | License | Author | Retrieved at |\n";
    content += "| --- | --- | --- | --- | --- | --- |\n";
  }

  for (const record of records) {
    content += buildSourceEntry({ ...record, retrievedAt });
  }

  await fs.promises.appendFile(sourceFile, content);
}

export async function run(argv = process.argv.slice(2)) {
  const parsed = parseArgs(argv);

  if (parsed.help) {
    console.log("[deprecated] Use the Python-first entry instead:");
    console.log("Usage: python3 scripts/fetch-guofeng-images.py --query \"West Lake Hangzhou\" \"Lingyin Temple Hangzhou\" --out site/assets/images --count 2 --width 1600");
    console.log("Legacy compatibility: node scripts/fetch-guofeng-images.mjs ...");
    return 0;
  }

  const out = path.resolve(parsed.out);
  const options = {
    ...parsed,
    out,
    siteRoot: siteRootForOut(out),
  };
  const sourceFile = sourcesPathForOut(out);
  const retrievedAt = new Date().toISOString().slice(0, 10);

  await fs.promises.mkdir(out, { recursive: true });

  const allRecords = [];
  for (const query of parsed.query) {
    try {
      const records = await fetchQuery(query, options);
      if (records.length === 0) {
        console.error(`[EMPTY] ${query}: Wikimedia Commons returned no downloadable images.`);
      } else {
        allRecords.push(...records);
        for (const record of records) {
          console.log(`[OK] ${record.localFile} | ${record.license} | ${record.artist} | ${record.sourceUrl}`);
        }
      }
    } catch (error) {
      console.error(`[FAIL] ${query}: ${error.message}`);
    }
  }

  if (allRecords.length === 0) {
    console.error("No images downloaded. Use browser/WebSearch, Unsplash/Pexels/manual sources, or honest placeholders with a material list.");
    return 1;
  }

  await appendSources(sourceFile, allRecords, retrievedAt);
  console.log(`\nDownloaded ${allRecords.length} image(s) to ${out}`);
  console.log(`Source log: ${path.relative(process.cwd(), sourceFile).replaceAll(path.sep, "/")}`);
  console.log("Review each image for 8/10 fit before using it in final HTML.");
  return 0;
}

const isCli = process.argv[1] && fileURLToPath(import.meta.url) === path.resolve(process.argv[1]);
if (isCli) {
  run().then((code) => {
    process.exitCode = code;
  }).catch((error) => {
    console.error(error.message);
    process.exitCode = 2;
  });
}

File: scripts/fetch-guofeng-images.py
#!/usr/bin/env python3

import json
import os
import re
import sys
from datetime import date
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode, urlparse
from urllib.request import Request, urlopen


API = "https://commons.wikimedia.org/w/api.php"
UA = "itbxm-guofeng-image-fetcher/1.0 (skill asset workflow)"


def parse_args(argv):
    parsed = {"query": [], "count": 2, "width": 1600}
    i = 0
    while i < len(argv):
        arg = argv[i]
        if arg == "--query":
            i += 1
            while i < len(argv) and not argv[i].startswith("--"):
                parsed["query"].append(argv[i])
                i += 1
            i -= 1
        elif arg == "--out":
            i += 1
            parsed["out"] = argv[i]
        elif arg == "--count":
            i += 1
            parsed["count"] = int(argv[i])
        elif arg == "--width":
            i += 1
            parsed["width"] = int(argv[i])
        elif arg in ("--help", "-h"):
            parsed["help"] = True
        else:
            raise ValueError(f"Unknown argument: {arg}")
        i += 1

    if not parsed.get("help"):
        if not parsed["query"]:
            raise ValueError("Missing --query")
        if not parsed.get("out"):
            raise ValueError("Missing --out")
        if parsed["count"] < 1:
            raise ValueError("--count must be a positive integer")
        if parsed["width"] < 200:
            raise ValueError("--width must be an integer >= 200")
    return parsed


def safe_file_name(value):
    value = re.sub(r"[^\w]+", "_", str(value), flags=re.UNICODE)
    value = re.sub(r"^_+|_+$", "", value)
    return value[:80] or "image"


def build_source_entry(local_file, query, source_url, license_name, artist, retrieved_at):
    return f"| {local_file} | {query} | {source_url} | {license_name or '?'} | {artist or '?'} | {retrieved_at} |\n"


def strip_html(value):
    value = re.sub(r"<[^>]+>", "", str(value or "?"))
    value = re.sub(r"\s+", " ", value).strip()
    return value or "?"


def image_extension(url):
    ext = Path(urlparse(url).path).suffix.lower()
    if ext in {".jpg", ".jpeg", ".png", ".webp", ".gif", ".svg"}:
        return ext
    return ".jpg"


def sources_path_for_out(out_dir):
    absolute_out = Path(out_dir).resolve()
    parent = absolute_out.parent
    sources_dir = parent / "sources" if parent.name == "assets" else parent.parent / "sources"
    return sources_dir / "image-sources.md"


def site_root_for_out(out_dir):
    absolute_out = Path(out_dir).resolve()
    parent = absolute_out.parent
    if parent.name == "assets":
        return parent.parent
    return Path.cwd()


def display_path(path):
    return os.path.relpath(path, Path.cwd()).replace(os.sep, "/")


def request_bytes(url, accept=None):
    headers = {"User-Agent": UA}
    if accept:
        headers["Accept"] = accept
    request = Request(url, headers=headers)
    with urlopen(request, timeout=30) as response:
        return response.read()


def api_get(params):
    url = f"{API}?{urlencode(params)}"
    return json.loads(request_bytes(url, "application/json").decode("utf-8"))


def download_file(url, file_path):
    file_path.write_bytes(request_bytes(url))


def fetch_query(query, options):
    data = api_get(
        {
            "action": "query",
            "format": "json",
            "generator": "search",
            "gsrsearch": query,
            "gsrnamespace": "6",
            "gsrlimit": str(options["count"]),
            "prop": "imageinfo",
            "iiprop": "url|extmetadata",
            "iiurlwidth": str(options["width"]),
        }
    )
    pages = list((data.get("query", {}).get("pages") or {}).values())
    records = []
    for page in pages[: options["count"]]:
        image_info = (page.get("imageinfo") or [{}])[0]
        image_url = image_info.get("thumburl") or image_info.get("url")
        if not image_url:
            continue

        meta = image_info.get("extmetadata") or {}
        license_name = strip_html((meta.get("LicenseShortName") or {}).get("value"))
        artist = strip_html((meta.get("Artist") or {}).get("value"))
        source_url = image_info.get("descriptionurl") or ""
        title = str(page.get("title") or "image").removeprefix("File:")
        file_name = f"{safe_file_name(f'{query}_{title}')}{image_extension(image_url)}"
        file_path = options["out"] / file_name
        download_file(image_url, file_path)
        records.append(
            {
                "local_file": file_path.relative_to(options["site_root"]).as_posix(),
                "query": query,
                "source_url": source_url,
                "license": license_name,
                "artist": artist,
            }
        )
    return records


def append_sources(source_file, records, retrieved_at):
    source_file.parent.mkdir(parents=True, exist_ok=True)
    content = ""
    if not source_file.exists():
        content += "# Image Sources\n\n"
        content += "| Local file | Query | Source URL | License | Author | Retrieved at |\n"
        content += "| --- | --- | --- | --- | --- | --- |\n"
    for record in records:
        content += build_source_entry(
            record["local_file"],
            record["query"],
            record["source_url"],
            record["license"],
            record["artist"],
            retrieved_at,
        )
    with source_file.open("a", encoding="utf-8") as handle:
        handle.write(content)


def run(argv):
    parsed = parse_args(argv)
    if parsed.get("help"):
        print(
            'Usage: python3 scripts/fetch-guofeng-images.py --query "West Lake Hangzhou" '
            '"Lingyin Temple Hangzhou" --out site/assets/images --count 2 --width 1600'
        )
        return 0

    out = Path(parsed["out"]).resolve()
    out.mkdir(parents=True, exist_ok=True)
    options = {**parsed, "out": out, "site_root": site_root_for_out(out)}
    source_file = sources_path_for_out(out)
    retrieved_at = date.today().isoformat()
    all_records = []

    for query in parsed["query"]:
        try:
            records = fetch_query(query, options)
            if not records:
                print(f"[EMPTY] {query}: Wikimedia Commons returned no downloadable images.", file=sys.stderr)
            else:
                all_records.extend(records)
                for record in records:
                    print(
                        f"[OK] {record['local_file']} | {record['license']} | "
                        f"{record['artist']} | {record['source_url']}"
                    )
        except (HTTPError, URLError, OSError, ValueError) as error:
            print(f"[FAIL] {query}: {error}", file=sys.stderr)

    if not all_records:
        print(
            "No images downloaded. Use browser/WebSearch, Unsplash/Pexels/manual sources, "
            "or honest placeholders with a material list.",
            file=sys.stderr,
        )
        return 1

    append_sources(source_file, all_records, retrieved_at)
    print(f"\nDownloaded {len(all_records)} image(s) to {out}")
    print(f"Source log: {display_path(source_file)}")
    print("Review each image for 8/10 fit before using it in final HTML.")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(run(sys.argv[1:]))
    except Exception as error:
        print(error, file=sys.stderr)
        sys.exit(2)
